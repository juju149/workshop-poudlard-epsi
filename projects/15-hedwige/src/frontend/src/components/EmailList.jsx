import React from 'react';
import '../styles/EmailList.css';

function EmailList({ emails, onEmailClick }) {
  if (!emails || emails.length === 0) {
    return (
      <div className="email-list-empty">
        <p>No emails found</p>
      </div>
    );
  }

  const formatDate = (dateString) => {
    const date = new Date(dateString);
    const now = new Date();
    const diffTime = Math.abs(now - date);
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

    if (diffDays === 0) {
      return date.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit' });
    } else if (diffDays === 1) {
      return 'Yesterday';
    } else if (diffDays < 7) {
      return date.toLocaleDateString('en-US', { weekday: 'short' });
    } else {
      return date.toLocaleDateString('en-US', { month: 'short', day: 'numeric' });
    }
  };

  const extractEmail = (from) => {
    const match = from.match(/<(.+)>/);
    return match ? match[1] : from;
  };

  const extractName = (from) => {
    const match = from.match(/^(.+?)\s*</);
    return match ? match[1].replace(/"/g, '') : from;
  };

  return (
    <div className="email-list">
      <h2>Inbox ({emails.length})</h2>
      <div className="email-items">
        {emails.map((email) => (
          <div 
            key={email.id} 
            className="email-item"
            onClick={() => onEmailClick(email.id)}
          >
            <div className="email-item-header">
              <span className="email-from">{extractName(email.from)}</span>
              <span className="email-date">{formatDate(email.date)}</span>
            </div>
            <div className="email-subject">{email.subject || '(No Subject)'}</div>
            <div className="email-snippet">{email.snippet}</div>
          </div>
        ))}
      </div>
    </div>
  );
}

export default EmailList;
