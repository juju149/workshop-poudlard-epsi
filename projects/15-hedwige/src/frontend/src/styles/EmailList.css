.email-list {
  padding: 20px;
}

.email-list h2 {
  margin-bottom: 20px;
  color: #333;
  font-size: 1.5em;
}

.email-items {
  display: flex;
  flex-direction: column;
  gap: 1px;
}

.email-item {
  padding: 16px 20px;
  border-bottom: 1px solid #e0e0e0;
  cursor: pointer;
  transition: background 0.2s;
}

.email-item:hover {
  background: #f9f9f9;
}

.email-item-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 6px;
}

.email-from {
  font-weight: 600;
  color: #333;
  font-size: 0.95em;
}

.email-date {
  color: #999;
  font-size: 0.85em;
}

.email-subject {
  font-weight: 500;
  color: #555;
  margin-bottom: 6px;
  font-size: 0.95em;
}

.email-snippet {
  color: #777;
  font-size: 0.9em;
  line-height: 1.4;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

.email-list-empty {
  padding: 60px 20px;
  text-align: center;
  color: #999;
  font-size: 1.1em;
}
