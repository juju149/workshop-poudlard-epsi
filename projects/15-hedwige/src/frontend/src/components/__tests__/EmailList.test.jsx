import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EmailList from '../EmailList';
import { vi } from 'vitest';

describe('EmailList', () => {
  const emails = [
    { id: 1, from: 'Harry Potter <harry@poudlard.fr>', subject: 'Bienvenue à Poudlard', date: '2025-10-10T10:00:00Z' },
    { id: 2, from: 'Hermione Granger <hermione@poudlard.fr>', subject: 'Devoirs', date: '2025-10-11T12:00:00Z' },
  ];

  it('affiche le message si la liste est vide', () => {
    render(<EmailList emails={[]} onEmailClick={() => {}} />);
    expect(screen.getByText(/No emails found/i)).toBeInTheDocument();
  });

  it('affiche la liste des emails', () => {
    render(<EmailList emails={emails} onEmailClick={() => {}} />);
    expect(screen.getByText(/Inbox \(2\)/i)).toBeInTheDocument();
    expect(screen.getByText(/Bienvenue à Poudlard/i)).toBeInTheDocument();
    expect(screen.getByText(/Devoirs/i)).toBeInTheDocument();
  });

  it('appelle onEmailClick lors du clic', () => {
    const handleClick = vi.fn();
    render(<EmailList emails={emails} onEmailClick={handleClick} />);
    fireEvent.click(screen.getByText(/Bienvenue à Poudlard/i));
    expect(handleClick).toHaveBeenCalledWith(1);
  });

  it('correspond au snapshot', () => {
    const { asFragment } = render(<EmailList emails={emails} onEmailClick={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

