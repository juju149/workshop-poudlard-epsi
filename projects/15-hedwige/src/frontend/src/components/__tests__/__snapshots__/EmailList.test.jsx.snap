// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailList > correspond au snapshot 1`] = `
<DocumentFragment>
  <div
    class="email-list"
  >
    <h2>
      Inbox (2)
    </h2>
    <div
      class="email-items"
    >
      <div
        class="email-item"
      >
        <div
          class="email-item-header"
        >
          <span
            class="email-from"
          >
            Harry Potter
          </span>
          <span
            class="email-date"
          >
            Fri
          </span>
        </div>
        <div
          class="email-subject"
        >
          Bienvenue Ã  Poudlard
        </div>
        <div
          class="email-snippet"
        />
      </div>
      <div
        class="email-item"
      >
        <div
          class="email-item-header"
        >
          <span
            class="email-from"
          >
            Hermione Granger
          </span>
          <span
            class="email-date"
          >
            Sat
          </span>
        </div>
        <div
          class="email-subject"
        >
          Devoirs
        </div>
        <div
          class="email-snippet"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
